src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioSolutionStatusServiceFactory.cs

```diff
@@ -0,0 +1,90 @@
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Experiments;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.LanguageServices.Experimentation;
+using Microsoft.VisualStudio.Shell;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation
+{
+    [ExportWorkspaceServiceFactory(typeof(ISolutionStatusService), ServiceLayer.Host), Shared]
+    internal class VisualStudioSolutionStatusServiceFactory : IWorkspaceServiceFactory
+    {
+        private readonly IThreadingContext _threadingContext;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public VisualStudioSolutionStatusServiceFactory(IThreadingContext threadingContext)
+        {
+            _threadingContext = threadingContext;
+        }
+
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+        {
+            if (workspaceServices.Workspace is VisualStudioWorkspace vsWorkspace)
+            {
+                var experimentationService = vsWorkspace.Services.GetService<IExperimentationService>();
+                if (!experimentationService.IsExperimentEnabled(WellKnownExperimentNames.PartialLoadMode))
+                {
+                    // don't enable partial load mode for ones that are not in experiement yet
+                    return SolutionStatusService.Default;
+                }
+
+                // only VSWorkspace supports partial load mode
+                return new Service(_threadingContext, vsWorkspace);
+            }
+
+            return SolutionStatusService.Default;
+        }
+
+        private class Service : ISolutionStatusService
+        {
+            private readonly IThreadingContext _threadingContext;
+            private readonly VisualStudioWorkspace _workspace;
+
+            public Service(IThreadingContext threadingContext, VisualStudioWorkspace workspace)
+            {
+                // until we get new platform API, use legacy one that is not fully do what we want
+                _threadingContext = threadingContext;
+                _workspace = workspace;
+            }
+
+            public async System.Threading.Tasks.Task WaitForAsync(Solution solution, CancellationToken cancellationToken)
+            {
+                if (_workspace.Options.GetOption(ExperimentationOptions.SolutionStatusService_ForceDelay))
+                {
+                    await System.Threading.Tasks.Task.Delay(_workspace.Options.GetOption(ExperimentationOptions.SolutionStatusService_DelayInMS)).ConfigureAwait(false);
+                }
+
+                var taskCompletionSource = new TaskCompletionSource<object>();
+
+                KnownUIContexts.SolutionExistsAndFullyLoadedContext.WhenActivated(() => taskCompletionSource.SetResult(null));
```

Comment:
this doesn't seem to have any relation to the Solution object passed in (probably why this API seems so weird).  It doesn't even use the object in any way.

Why is this not just a WorkspaceService that you simply await on instead?

Reply:
I already wrote in the PR saying this is heuristic until real platform API is provided. it uses feature flag so it will no op for people who are not included in feature ring.

Reply:
> I already wrote in the PR saying this is heuristic until real platform API is provided. it uses feature flag so it will no op for people who are not included in feature ring.

This doesn't address the question i raised.  

Even as heuristic, it doesn't use or need Solution.  If, the final API is somehow related to a specific solution (not sure what that would mean, but ok), it doesn't make sense to still ask to `WaitFor(Solution)`.  A more sensible API would instead be `WaitFor(SolutionId)` indicating that you're referencing the abstract concept of this Solution and waiting that abstract concept to reach a fully loaded point.   Once that was done, you could go grab the actual 'workspace.CurrentSolution' which would be the appropriate Solution to now use.

--

In other words, you're mixing two concepts that don't make sense.  You have the concept of the Solution (which is immutable and cannot be obvserved to change) with the concept of "waiting for a certain poitn in time so that all the data is ready".  These two guys dont' mix.  Once is about waiting for something to be up to date.  The ohter is about being fixed and never changing.  

I am proposing that your operators work on Roslyn objects that better convey the concepts you're trying to combine.  For example, you could say "i want to wait for this workspace to reach some reasonable point" since hte Workspace represent a mutable concept.  Similarly, you could say "i want to wait until there is a reasonable solution snapshot for this given solution id".

Reply:
sure. we can go to solutoinId. but I am starting from Solution, Project since projectid doesn't have projectId.SolutionId back pointer. only Project has Project.Solution back pointer.

that is just small thing to tweak. not related to the main point it is trying to achieve.

if that was your concern you could just suggest, how about using solutionId rather than solution.

Reply:
Is there a reason you need the SolutionId?  It's unclear to me what even *that* would be used for.

Reply:
we probably will start with whether a solution is fully loaded, but we want to be more granular than that. like whether current project and its dependency is fully loaded or whether current project's sources are all ready and etc.

we don't know exact shape of it yet. if our projectId had back pointer to solutionId, I would have used it.

Reply:
> we probably will start with whether a solution is fully loaded, but we want to be more granular than that. like whether current project and its dependency is fully loaded or whether current project's sources are all ready and etc.

So... it seems like it's sufficient to just have overloads for SolutionId and ProjectId.   Why is it necessary for ProjectId to point back to SolutionId?